File: src/main/java/com/example/leaderboard/domain/ranking/service/RankingService.java
package com.example.leaderboard.domain.ranking.service;

import com.example.leaderboard.domain.ranking.dto.RankingEntry;
import com.example.leaderboard.domain.ranking.dto.RankingResponse;
import com.example.leaderboard.domain.ranking.repository.RankingRepository;
import com.example.leaderboard.domain.user.entity.User;
import com.example.leaderboard.domain.user.repository.UserRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.data.redis.core.RedisTemplate;
import org.springframework.data.redis.core.ZSetOperations;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.Set;
import java.util.stream.Collectors;

@Service
@Transactional(readOnly = true)
@RequiredArgsConstructor
public class RankingService {

    private final RankingRepository rankingRepository;
    private final UserRepository userRepository;
    private final RedisTemplate<String, String> redisTemplate;

    public RankingResponse getRanking(String puzzleId, int limit) {
        String zsetKey = "ranking:" + puzzleId;
        Set<ZSetOperations.TypedTuple<String>> tuples =
                redisTemplate.opsForZSet().reverseRangeWithScores(zsetKey, 0, limit - 1L);

        // Redis에 데이터 없으면 DB fallback
        if (tuples == null || tuples.isEmpty()) {
            return new RankingResponse(puzzleId, rankingRepository.findTopRankingByPuzzleId(puzzleId, limit));
        }

        List<Long> userIds = tuples.stream()
                .filter(t -> t.getValue() != null)
                .map(t -> Long.parseLong(t.getValue()))
                .toList();

        Map<Long, String> userMap = userRepository.findAllByIdIn(userIds).stream()
                .collect(Collectors.toMap(User::getId, User::getUsername));

        List<RankingEntry> entries = new ArrayList<>();
        int rank = 1;
        for (ZSetOperations.TypedTuple<String> tuple : tuples) {
            if (tuple.getValue() == null || tuple.getScore() == null) continue;
            Long userId = Long.parseLong(tuple.getValue());
            String username = userMap.get(userId);
            if (username == null) continue;
            entries.add(new RankingEntry(rank++, userId, username, tuple.getScore().longValue()));
        }

        return new RankingResponse(puzzleId, entries);
    }
}

File: src/main/kotlin/com/example/leaderboard/domain/ranking/service/RankingService.kt
package com.example.leaderboard.domain.ranking.service

import com.example.leaderboard.domain.ranking.dto.RankingEntry
import com.example.leaderboard.domain.ranking.dto.RankingResponse
import com.example.leaderboard.domain.ranking.repository.RankingRepository
import com.example.leaderboard.domain.user.repository.UserRepository
import org.springframework.data.redis.core.RedisTemplate
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional

@Service
@Transactional(readOnly = true)
class RankingService(
    private val rankingRepository: RankingRepository,
    private val userRepository: UserRepository,
    private val redisTemplate: RedisTemplate<String, String>,
) {

    fun getRanking(puzzleId: String, limit: Int = 100): RankingResponse {
        val zsetKey = "ranking:$puzzleId"
        val tuples = redisTemplate.opsForZSet()
            .reverseRangeWithScores(zsetKey, 0, limit - 1L)

        // Redis ZSET에 데이터가 없으면 DB 풀스캔으로 fallback
        if (tuples.isNullOrEmpty()) {
            return RankingResponse(
                puzzleId = puzzleId,
                entries = rankingRepository.findTopRankingByPuzzleId(puzzleId, limit),
            )
        }

        val userIds = tuples.mapNotNull { it.value?.toLongOrNull() }
        val userMap = userRepository.findAllByIdIn(userIds).associateBy { it.id }

        val entries = tuples.mapIndexedNotNull { index, tuple ->
            val userId = tuple.value?.toLongOrNull() ?: return@mapIndexedNotNull null
            val username = userMap[userId]?.username ?: return@mapIndexedNotNull null
            RankingEntry(
                rank = index + 1,
                userId = userId,
                username = username,
                totalScore = tuple.score?.toLong() ?: 0L,
            )
        }

        return RankingResponse(puzzleId = puzzleId, entries = entries)
    }
}
